spacelift/resource_security_email.go

```diff
@@ -0,0 +1,87 @@
+package spacelift
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/spacelift-io/terraform-provider-spacelift/spacelift/internal"
+)
+
+func resourceSecurityEmail() *schema.Resource {
+	return &schema.Resource{
+		Description: "" +
+			"`spacelift_security_email` represents a security email address that " +
+			"receives notifications about security issues in Spacelift.",
+
+		CreateContext: resourceSecurityEmailCreate,
+		ReadContext:   resourceSecurityEmailRead,
+		UpdateContext: resourceSecurityEmailUpdate,
+		DeleteContext: resourceSecurityEmailDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"email": {
+				Type:        schema.TypeString,
+				Description: "Email address to which the security notifications are sent",
+				Required:    true,
+			},
+		},
+	}
+}
+
+func resourceSecurityEmailCreate(ctx context.Context, data *schema.ResourceData, i interface{}) diag.Diagnostics {
+	var mutation struct {
+		SecurityEmail *string `graphql:"accountUpdateSecurityEmail(securityEmail: $securityEmail)"`
+	}
+
+	variables := map[string]interface{}{"securityEmail": toString(data.Get("email"))}
+	if err := i.(*internal.Client).Mutate(ctx, "AccountUpdateSecurityEmail", &mutation, variables); err != nil {
+		return diag.Errorf("could not create security email: %v", err)
+	}
+
+	data.SetId(time.Now().String())
+
+	return resourceSecurityEmailRead(ctx, data, i)
+}
+
+func resourceSecurityEmailRead(ctx context.Context, data *schema.ResourceData, i interface{}) diag.Diagnostics {
+	var query struct {
+		SecurityEmail *string `graphql:"securityEmail"`
+	}
+	if err := i.(*internal.Client).Query(ctx, "SecurityEmail", &query, nil); err != nil {
+		return diag.Errorf("could not query for security email: %v", err)
+	}
+
+	if query.SecurityEmail == nil {
+		data.SetId("")
+		return nil
+	}
+
+	data.Set("email", query.SecurityEmail)
+
+	return nil
+}
+
+func resourceSecurityEmailUpdate(ctx context.Context, data *schema.ResourceData, i interface{}) diag.Diagnostics {
+	var mutation struct {
+		SecurityEmail *string `graphql:"accountUpdateSecurityEmail(securityEmail: $email)"`
+	}
+	variables := map[string]interface{}{
+		"email": toString(data.Get("email")),
+	}
+	if err := i.(*internal.Client).Mutate(ctx, "AccountUpdateSecurityEmail", &mutation, variables); err != nil {
+		return diag.Errorf("could not create security email: %v", err)
+	}
+
+	return resourceSecurityEmailRead(ctx, data, i)
+}
+
+func resourceSecurityEmailDelete(ctx context.Context, data *schema.ResourceData, i interface{}) diag.Diagnostics {
+	return diag.Errorf("deleting security email is not supported")
```

Comment:
This doesn't seem a great UX. Wouldn't it mean that if you had added this resource, you'd need to remove it from the Terraform state in order to be able to delete the resource?

I wonder if we should do something like this:

1. Update the API to allow the security email to be set to an empty string / null; or
2. Turn this into a no-op (maybe with a warning).

Reply:
> you'd need to remove it from the Terraform state in order to be able to delete the resource?

yes

I'll turn it into a warning and remove the resource from the state